File: backend/Application/Services/SubWorkItemService.cs
using backend.Application.Interfaces;

using backend.Application.Exceptions;
namespace backend.Application.Services;

public class SubWorkItemService : ISubWorkItemService
{
    private readonly ISubWorkItemRepository _subRepo;
    private readonly IWorkItemRepository _workItemRepo;
    private readonly IBoardRepository _boardRepo;
    private readonly IWorkspaceMemberRepository _workspaceMemberRepo;
    private readonly IBoardAccessService _boardAccessService;
    private readonly IWorkItemLockService _workItemLockService;
    private readonly IUserContext _userContext;
    private readonly DomainEventSubject _domainEventSubject;
    private readonly CommandInvoker _commandInvoker;

    public SubWorkItemService(
        ISubWorkItemRepository subRepo,
        IWorkItemRepository workItemRepo,
        IBoardRepository boardRepo,
        IWorkspaceMemberRepository workspaceMemberRepo,
        IBoardAccessService boardAccessService,
        IWorkItemLockService workItemLockService,
        IUserContext userContext,
        DomainEventSubject domainEventSubject,
        CommandInvoker commandInvoker)
    {
        _subRepo = subRepo;
        _workItemRepo = workItemRepo;
        _boardRepo = boardRepo;
        _workspaceMemberRepo = workspaceMemberRepo;
        _boardAccessService = boardAccessService;
        _workItemLockService = workItemLockService;
        _userContext = userContext;
        _domainEventSubject = domainEventSubject;
        _commandInvoker = commandInvoker;
    }

    public async Task<SubWorkItemResponse> Create(SubWorkItemRequest request)
    {
        var command = new CreateSubWorkItemCommand(
            request,
            _userContext.GetUserId(),
            _subRepo,
            _workItemRepo,
            _boardRepo,
            _workspaceMemberRepo,
            _boardAccessService,
            _workItemLockService,
            _domainEventSubject);

        return await _commandInvoker.ExecuteAsync(command);
    }

    public async Task<SubWorkItemResponse> GetById(int subWorkItemId)
    {
        var subWorkItem = await _subRepo.GetById(subWorkItemId);
        if (subWorkItem == null)
            throw new NotFoundException("Sub work item does not exist.");

        var workItem = await _workItemRepo.GetById(subWorkItem.WorkItemId);
        if (workItem == null)
            throw new NotFoundException("Work item does not exist.");

        var board = await _boardRepo.GetById(workItem.BoardId);
        if (board == null)
            throw new NotFoundException("Board does not exist.");

        await _boardAccessService.EnsureBoardAccessAsync(board);

        return SubWorkItemHelper.ToResponse(subWorkItem);
    }

    public async Task<List<SubWorkItemResponse>> GetByWorkItemId(int workItemId)
    {
        var workItem = await _workItemRepo.GetById(workItemId);
        if (workItem == null)
            throw new NotFoundException("Work item does not exist.");

        var board = await _boardRepo.GetById(workItem.BoardId);
        if (board == null)
            throw new NotFoundException("Board does not exist.");

        await _boardAccessService.EnsureBoardAccessAsync(board);

        var subItems = await _subRepo.GetByWorkItemIdAsync(workItemId);

        return subItems.Select(SubWorkItemHelper.ToResponse).ToList();
    }

    public async Task<SubWorkItemResponse> Update(
        int subWorkItemId,
        UpdateSubWorkItemRequest request)
    {
        var command = new UpdateSubWorkItemCommand(
            subWorkItemId,
            request,
            _userContext.GetUserId(),
            _subRepo,
            _workItemRepo,
            _boardRepo,
            _workspaceMemberRepo,
            _boardAccessService,
            _workItemLockService,
            _domainEventSubject);

        return await _commandInvoker.ExecuteAsync(command);
    }

    public async Task<SubWorkItemResponse> ChangeStatus(
        int subWorkItemId,
        ChangeSubWorkItemStatusRequest request)
    {
        var command = new ChangeSubWorkItemStatusCommand(
            subWorkItemId,
            request,
            _userContext.GetUserId(),
            _subRepo,
            _workItemRepo,
            _boardRepo,
            _boardAccessService,
            _workItemLockService,
            _domainEventSubject);

        return await _commandInvoker.ExecuteAsync(command);
    }

    public async Task Delete(int subWorkItemId)
    {
        var command = new DeleteSubWorkItemCommand(
            subWorkItemId,
            _userContext.GetUserId(),
            _subRepo,
            _workItemRepo,
            _boardRepo,
            _boardAccessService,
            _workItemLockService,
            _domainEventSubject);

        await _commandInvoker.ExecuteAsync(command);
    }
}

File: frontend/shared/services/sub-work-item.service.ts
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import {
  ChangeSubWorkItemStatusRequest,
  SubWorkItemRequest,
  SubWorkItemResponse,
  UpdateSubWorkItemRequest,
} from '../interfaces';
import { TaskStatus } from '../types/task-status.type';

@Injectable({ providedIn: 'root' })
export class SubWorkItemService {
  private readonly apiUrl = 'http://localhost:8080/api/sub-work-items';

  constructor(private readonly http: HttpClient) {}

  private authHeaders(): HttpHeaders {
    const token = localStorage.getItem('token');
    return new HttpHeaders({
      Authorization: `Bearer ${token ?? ''}`,
    });
  }

  public getById(subWorkItemId: number) {
    return this.http.get<SubWorkItemResponse>(`${this.apiUrl}/${subWorkItemId}`, {
      headers: this.authHeaders(),
    });
  }

  public getByWorkItemId(workItemId: number) {
    return this.http.get<SubWorkItemResponse[]>(
      `${this.apiUrl}/work-item/${workItemId}`,
      { headers: this.authHeaders() }
    );
  }

  public create(request: SubWorkItemRequest) {
    return this.http.post<SubWorkItemResponse>(this.apiUrl, request, {
      headers: this.authHeaders(),
    });
  }

  public update(subWorkItemId: number, request: UpdateSubWorkItemRequest) {
    return this.http.put<SubWorkItemResponse>(
      `${this.apiUrl}/${subWorkItemId}`,
      request,
      { headers: this.authHeaders() }
    );
  }

  public changeStatus(subWorkItemId: number, status: TaskStatus) {
    const payload: ChangeSubWorkItemStatusRequest = { status };
    return this.http.patch<SubWorkItemResponse>(
      `${this.apiUrl}/${subWorkItemId}/status`,
      payload,
      { headers: this.authHeaders() }
    );
  }

  public delete(subWorkItemId: number) {
    return this.http.delete<void>(`${this.apiUrl}/${subWorkItemId}`, {
      headers: this.authHeaders(),
    });
  }
}
